src/backend/distributed/planner/query_colocation_checker.c

```diff
@@ -194,6 +194,19 @@ SubqueryColocated(Query *subquery, ColocatedJoinChecker *checker)
 	RelationRestrictionContext *unionedRelationRestrictionContext = NULL;
 	PlannerRestrictionContext *unionedPlannerRestrictionContext = NULL;
 
+	/*
+	 * There are no relations in the input subquery, such as a subquery
+	 * that consist of only intermediate results. At this point of planning,
+	 * where we're evaluating non-colocated subqueries (e.g., individual
+	 * subqueries have already been recursively planned), we're sure that
+	 * this can only be a query that consist of only a single intermediate
+	 * result.
+	 */
+	if (list_length(filteredRestrictionList) == 0)
+	{
```

Comment:
One idea could be to keep `QueryAlreadyRecursivelyPlanned()` (or probably something more generic that covers subqueries with intermediate results only, e.g., including subqueries as well) and add an assert here before returning.

Do you think would it worth adding it?

Reply:
Is that the only case we might see here?

Reply:
As far as I can see, yes.

First, since the `filteredRestrictionList` is `NULL`, we know that there are no tables in the subquery. If it were, we'd have entry for it.

Also, given that we're in non-colocated subquery checks, all the other individual subqueries that does not include relations (e.g., values, functions, ctes) have already been recursively planned. 

So, I cannot think of a case at this point where we're dealing with a subquery that doesn't have a relation in it and it is not an intermediate result. My point for assertion is that in case we change the order of processing.

Reply:
Can't we have other subqueries without tables? (e.g. SELECT 1) ?

Reply:
My point is that those subqueries, which are in the same level of the subquery we're looking at this point, should have already been recursively planned. Am I missing something?

Reply:
Ok, so I see what you mean now, the following query is an example: `SELECT count(*) FROM users_table, (SELECT 2 ) as foo(x)  WHERE foo.x = users_table.value_1 AND  user_id IN (SELECT user_id+1 FROM events_table);` 

So, we cannot do the assertion as I suggested. But, the check still sounds good to me. If there is no distributed table**, there is no way for non-colocated subquery joins.  

What do you think?

** reference table are handled below in `RestrictionEquivalenceForPartitionKeysViaEquivalances `

Reply:
In the following example, we should not be recursively planning `SELECT 1`.  If we do that's a bug. But SELECT 1 would not match the assert.
```
SELECT *
FROM
  (SELECT 1 AS user_id) AS users_table
JOIN LATERAL
  (SELECT *
   FROM
     (SELECT *
      FROM events_table
      WHERE user_id = users_table.user_id) AS bar
   LEFT JOIN users_table u2 ON u2.user_id = bar.value_2) AS foo ON TRUE
```

Reply:
> So, we cannot do the assertion as I suggested. But, the check still sounds good to me. If there is no distributed table**, there is no way for non-colocated subquery joins.

oh yes, the check is fine. :)